nsiqcppstyle_rulemanager.py

```diff
@@ -180,62 +200,154 @@ def ResetRegisteredRules(self):
         del self.typeScopeRules[:]
         del self.fileStartRules[:]
         del self.fileEndRules[:]
+        del self.sessionStartRules[:]
+        del self.sessionEndRules[:]
         del self.projectRules[:]
         del self.preprocessRules[:]
+        del self.commentRules[:]
 
     def AddPreprocessRule(self, preprocessRule):
-        """ Add rule which runs in preprocess statements """
+        """
+        Add rule which runs in preprocess statements
+        Added rule should be function with following prototype:
+            def RunRule(lexer, contextStack)
+                lexer is the lexer used to analyze the source, it points the start token of source.
+                contextStack is a stack showing the lexical context the token resides in
+        """
         self.preprocessRules.append(preprocessRule)

```

Comment:
Instead of directly adding the rule, we could process it to make the code in `RunXRules` simpler:

```py
from contextlib import ContextDecorator

class LexerContext(ContextDecorator):
    _lexer = None
    def __enter__(self):
        data = self._lexer.Backup()
        return self

    def __exit__(self, *exc):
        self._lexer.Restore(data)
        return False

    def __call__(self, func):
        safe_func = super().__call__(func)
        @wraps(safe_func)
        def inner(lexer, *args, **kwds):
            self._lexer = lexer
            return safe_func(lexer, *args, **kwds)
        return inner
```

With this context manager, we can do:
```py
        context = LexerContext()
        self.preprocessRules.append(context(preprocessRule))
```

And then the `RunXRules` becomes just a loop and calling the relevant functions from the dict, just like other non-lexer based functions (so we can make just 1 function called by all user-facing functions)

TBH, I don't know if this would work, but it provides us with a single place to:
* add features regarding lexer callbacks in future
* handle errors in user-provided functions (try-except to continue even when user function fails)

Reply:
Very neat!  Working with you for a relatively short amount of time has already sharpened my understanding of finer Python features such as types, context managers, decorators, and more - many thanks, and looking forward to learning more!